test/fuzz/transformation_record_synonymous_constants_test.cpp

```diff
@@ -0,0 +1,286 @@
+      context.get(), transformation_context));
+
+  // Apply transformation
+  TransformationRecordSynonymousConstants(9, 11).Apply(context.get(),
+                                                       &transformation_context);
+  // Check that the two ids are now synonyms
+  ASSERT_TRUE(transformation_context.GetFactManager()->IsSynonymous(
+      MakeDataDescriptor(9, {}), MakeDataDescriptor(11, {})));
+}
+
+TEST(TransformationRecordSynonymousConstantsTest, BoolConstants) {
+  std::string shader = R"(
+               OpCapability Shader
+          %1 = OpExtInstImport "GLSL.std.450"
+               OpMemoryModel Logical GLSL450
+               OpEntryPoint Fragment %4 "main" %19
+               OpExecutionMode %4 OriginUpperLeft
+               OpSource ESSL 310
+               OpName %4 "main"
+               OpName %8 "b"
+               OpName %19 "color"
+               OpDecorate %19 Location 0
+          %2 = OpTypeVoid
+          %3 = OpTypeFunction %2
+          %6 = OpTypeBool
+          %7 = OpTypePointer Function %6
+          %9 = OpConstantFalse %6
+         %20 = OpConstantNull %6
+         %11 = OpConstantTrue %6
+         %16 = OpTypeFloat 32
+         %17 = OpTypeVector %16 4
+         %18 = OpTypePointer Output %17
+         %19 = OpVariable %18 Output
+          %4 = OpFunction %2 None %3
+          %5 = OpLabel
+          %8 = OpVariable %7 Function
+               OpStore %8 %9
+         %10 = OpLoad %6 %8
+         %12 = OpLogicalEqual %6 %10 %11
+               OpSelectionMerge %14 None
+               OpBranchConditional %12 %13 %14
+         %13 = OpLabel
+               OpReturn
+         %14 = OpLabel
+               OpReturn
+               OpFunctionEnd
+  )";
+
+  const auto env = SPV_ENV_UNIVERSAL_1_4;
+  const auto consumer = nullptr;
+  const auto context = BuildModule(env, consumer, shader, kFuzzAssembleOption);
+
+  FactManager fact_manager;
+  spvtools::ValidatorOptions validator_options;
+  TransformationContext transformation_context(&fact_manager,
+                                               validator_options);
+  ASSERT_TRUE(IsValid(env, context.get()));
+
+  // %9 and %11 are not equivalent
+  ASSERT_FALSE(TransformationRecordSynonymousConstants(9, 11).IsApplicable(
+      context.get(), transformation_context));
+
+  // %20 and %11 are not equivalent
+  ASSERT_FALSE(TransformationRecordSynonymousConstants(20, 11).IsApplicable(
+      context.get(), transformation_context));
+
+  // %9 and %20 are equivalent
+  ASSERT_TRUE(TransformationRecordSynonymousConstants(9, 20).IsApplicable(
+      context.get(), transformation_context));
+
+  // Apply transformation
+  TransformationRecordSynonymousConstants(9, 20).Apply(context.get(),
+                                                       &transformation_context);
+
+  // Check that the ids are now synonyms
+  transformation_context.GetFactManager()->IsSynonymous(
+      MakeDataDescriptor(9, {}), MakeDataDescriptor(20, {}));
```

Comment:
Can you enhance the test to check that:

- if there are multiple OpConstantTrues they can be regarded as synonymous
- Same for multiple OpConstantFalses?

Reply:
Done